### ISSUE TITLE
[UX] - Help user of insert link dialogue by automatically adding URL protocol prefix to external links.

### ISSUE THREAD

**Issue by Graham-72** [backdrop/backdrop-issues#3083]
When using the WYSIWYG editor to add a link to an external URL it is necessary to enter the full URL including the protocol and this needs to be made clear to the user e.g. http://backdropcms.org.

**The PR is now modified so that the URL protocol prefix is added automatically to external links.**

~~I suggest that an additional line of description is added, such as `External paths (URL) must include a protocol prefix e.g. http:// .` as in the screenshot below.~~
![screenshot-2018-5-6 edit basic page a new test page backdrop starter](https://user-images.githubusercontent.com/10559274/39672172-0387d22c-511d-11e8-9cf2-9580f3f86448.png)

**Comment by docwilmot** [comment]
I'd be in favour of parsing the supplied url to determine if it is absolute or relative and adding the protocol if it is missing.

**Comment by Graham-72** [comment]
That would be good. Would one assume a default protocol of http:// ?

**Comment by olafgrabienski** [comment]
> Would one assume a default protocol of http:// ?

I'd say yes, HTTP (and HTTPS) are the most expectable protocols in an external link. If an external site uses HTTPS, it usually provides - or at least it should - a redirect from HTTP to HTTPS.

If someone wants to link to another protocol like FTP, I guess he or she would do that by his/herself. Btw, we have to make sure that `http://` doesn't get added to FTP, *mailto* or other protocols different from HTTP(S).

**Comment by docwilmot** [comment]
Since this downloads then links a file as you describe, doesn't it error if the file is not found?

**Comment by docwilmot** [comment]
I've seen some editors use a dropdown for available protocols, with http as default.

**Comment by olafgrabienski** [comment]
> perhaps it would help the user to add the protocol if the user enters a typical URL starting with 'www.'

That would help a bit but http URLs without 'www' are also very common, which makes the recognition of a typical URL pattern more complicated.

**Comment by olafgrabienski** [comment]
In the meantime I've had a look at the link field to see how it works: It adds the HTTP protocol.

If you put type for instance `www.example.com` in the URL part of the field, the displayed link URL becomes `http://www.example.com`. It works also without www: `example.com` becomes `http://example.com`.

FTP is not treated differently, it also just gets the `http://` part.

Don't know the technical details, but as a first step I'd suggest to imitate the link field behaviour for consistency.

**Comment by Graham-72** [comment]
@olafgrabienski thanks for this good suggestion. The links module in core looks complicated and it may take a while to work out what parts are needed here. In the mean time I would like to see the simple fix in my PR applied as soon as possible as a help for users.

**Comment by olafgrabienski** [comment]
> In the mean time I would like to see the simple fix in my PR applied as soon as possible as a help for users.

I've had a look at the sandbox site. This is the help text:

> External paths (URL) must include a protocol prefix e.g. http:// .

Not sure: do we need a comma between "prefix" and "e.g."? And it is possible to display `http://` as code? That would allow to omit the space before the period:

> External paths (URL) must include a protocol prefix, for instance `http://`.

**Comment by docwilmot** [comment]
> In the meantime I've had a look at the link field to see how it works: It adds the HTTP protocol.

Good call. If you run the user entered URL through `link_cleanup_url()` (or copy the function if Link isnt always enabled) then this function validates that its an external URL and slaps on an `http://` for you if its missing. Simplest solution so far.

**Comment by Graham-72** [comment]
> do we need a comma between "prefix" and "e.g."? And it is possible to display http:// as code? 

I have modified the PR to do this.

> If you run the user entered URL through link_cleanup_url()

I will try adding this, hopefully in the near future.

**Comment by quicksketch** [comment]
Link module's auto-http prefixing code is indeed a bit complicated. I don't think we could easily copy it into another module or into core's common.inc. However, most sites will have it enabled. Perhaps auto-clean-up on submit using it, and then if it *still* fails throw the error to the user as a validation error?

I also would like to avoid filling up description text, it would definitely be preferable to automatically fix the problem if possible.

**Comment by Graham-72** [comment]
> it would definitely be preferable to automatically fix the problem.

I have revised the PR so that instead of providing advice to the user as previously proposed, the procedure now uses  `function link_validate_url()` from the link module together with a slightly adapted version of `function link_cleanup_url()`, copied and modified here as `function _filter_cleanup_url()`. The modification is in order that the function returns false if the url is not found to be external.

Thanks also to @docwilmot and @olafgrabienski for the suggestion.

**Comment by olafgrabienski** [comment]
@Graham-72 I've tested the PR in the sandbox site:

- `www.example.com` becomes `http://www.example.com`
- `example.com` becomes `http://example.com`
- `/about` remains `/about`

Looks good!

**Comment by herbdool** [comment]
@quicksketch I've tested and it works. I get a warning if link is disabled and link is relative without the http.

**Comment by quicksketch** [comment]
Thanks @Graham-72 and @olafgrabienski! I tested the PR and it works well. I'm not a fan of copy/pasting all those complicated regular expressions out of Link module though. I've made a new PR at https://github.com/backdrop/backdrop/pull/2188 that avoids duplicating any code (though it is slightly less efficient because `link_validate_url()` is called twice).

I also found that non-URL links like `mailto:foo@example.com` had been broken by the leading slash check. Rather than throw a validation error, my update auto-prefixes the leading slash. Then we can keep the URL-checking the same as before, just the clean-up is done automatically if Link module exists.
